clash-lib/src/Clash/Normalize/PrimitiveReductions.hs

```diff
@@ -71,6 +79,183 @@ import           Clash.Rewrite.Types
 import           Clash.Rewrite.Util
 import           Clash.Unique
 import           Clash.Util
+import qualified Clash.Util.Interpolate           as I
+
+typeNatAdd :: TyConName
+typeNatAdd =
+  Name User "GHC.TypeNats.+" (getKey typeNatAddTyFamNameKey) wiredInSrcSpan
+
+typeNatMul :: TyConName
+typeNatMul =
+  Name User "GHC.TypeNats.*" (getKey typeNatMulTyFamNameKey) wiredInSrcSpan
+
+typeNatSub :: TyConName
+typeNatSub =
+  Name User "GHC.TypeNats.-" (getKey typeNatSubTyFamNameKey) wiredInSrcSpan
+
+vecHeadPrim
+  :: TyConName
+  -- ^ Vec TyCon name
+  -> Term
+vecHeadPrim vecTcNm =
+ -- head :: Vec (n+1) a -> a
+  Prim (PrimInfo "Clash.Sized.Vector.head" (vecHeadTy vecTcNm) WorkNever)
+
+vecLastPrim
+  :: TyConName
+  -- ^ Vec TyCon name
+  -> Term
+vecLastPrim vecTcNm =
+  -- last :: Vec (n+1) a -> a
+  -- has the same type signature as head, hence we're reusing its type
+  -- definition here.
+  Prim (PrimInfo "Clash.Sized.Vector.last" (vecHeadTy vecTcNm) WorkNever)
+
+vecHeadTy
+  :: TyConName
+  -- ^ Vec TyCon name
+  -> Type
+vecHeadTy vecNm =
+  ForAllTy nTV $
+  ForAllTy aTV $
+  mkFunTy
+    (mkTyConApp vecNm [mkTyConApp typeNatAdd [VarTy nTV, LitTy (NumTy 1)], VarTy aTV])
+    (VarTy aTV)
+ where
+  aTV = mkTyVar liftedTypeKind (mkUnsafeSystemName "a" 0)
+  nTV = mkTyVar typeNatKind (mkUnsafeSystemName "n" 1)
+
+vecTailPrim
+  :: TyConName
+  -- ^ Vec TyCon name
+  -> Term
+vecTailPrim vecTcNm =
+  -- tail :: Vec (n + 1) a -> Vec n a
+  Prim (PrimInfo "Clash.Sized.Vector.tail" (vecTailTy vecTcNm) WorkNever)
+
+vecInitPrim
+  :: TyConName
+  -- ^ Vec TyCon name
+  -> Term
+vecInitPrim vecTcNm =
+  -- init :: Vec (n + 1) a -> Vec n a
+  -- has the same type signature as tail, hence we're reusing its type
+  -- definition here.
+  Prim (PrimInfo "Clash.Sized.Vector.init" (vecTailTy vecTcNm) WorkNever)
+
+vecTailTy
+  :: TyConName
+  -- ^ Vec TyCon name
+  -> Type
+vecTailTy vecNm =
+  ForAllTy nTV $
+  ForAllTy aTV $
+  mkFunTy
+    (mkTyConApp vecNm [mkTyConApp typeNatAdd [VarTy nTV, LitTy (NumTy 1)], VarTy aTV])
+    (mkTyConApp vecNm [VarTy nTV, VarTy aTV])
+ where
+  nTV = mkTyVar typeNatKind (mkUnsafeSystemName "n" 0)
+  aTV = mkTyVar liftedTypeKind (mkUnsafeSystemName "a" 1)
+
+-- | Makes two case statements: the first one extract the _head_ from the given
+-- vector, the latter the tail.
+extractHeadTail
+  :: DataCon
+  -- ^ The Cons (:>) constructor
+  -> Type
+  -- ^ Element type
+  -> Integer
+  -- ^ Length of the vector
+  -> Term
+  -- ^ Vector to extract head from
+  -> (Term, Term)
+  -- ^ (head of vector, tail of vector)
+extractHeadTail consCon elTy n vec =
+  ( Case vec elTy [(pat, Var el)]
+  , Case vec restTy [(pat, Var rest)] )
```

Comment:
@martijnbastiaan  Doesn't this duplicate (at least compiler) work for the `vec` term?

Reply:
Possibly / probably. How would you write it?

There should also be a "quick" path that just extract the element if it sees a data constructor.